File: backend/src/main/kotlin/com/alpha/domain/repository/CategoryRepository.kt
package com.alpha.domain.repository

import com.alpha.domain.entity.CategoryEntity
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.query.Param
import java.util.*

interface CategoryRepository : JpaRepository<CategoryEntity, UUID> {
    fun findBySlug(slug: String): CategoryEntity?
    @Query("SELECT c FROM CategoryEntity c WHERE c.parent.id = :parentId")
    fun findByParentId(@Param("parentId") parentId: UUID?): List<CategoryEntity>
    
    @Query("SELECT c FROM CategoryEntity c WHERE c.isActive = true")
    fun findByIsActiveTrue(): List<CategoryEntity>
    
    fun existsBySlug(slug: String): Boolean
    
    @Query("SELECT c FROM CategoryEntity c WHERE c.parent IS NULL ORDER BY c.sortOrder")
    fun findRootCategories(): List<CategoryEntity>
    
    @Query("SELECT c FROM CategoryEntity c WHERE LOWER(c.name) LIKE LOWER(CONCAT('%', :search, '%')) OR LOWER(c.slug) LIKE LOWER(CONCAT('%', :search, '%'))")
    fun search(@Param("search") search: String): List<CategoryEntity>
    
    @Query("SELECT c FROM CategoryEntity c WHERE c.parent.id = :parentId ORDER BY c.sortOrder")
    fun findByParentIdOrdered(@Param("parentId") parentId: UUID): List<CategoryEntity>
}

File: backend-scala-zio/src/main/scala/com/alpha/repository/CategoryRepository.scala
package com.alpha.repository

import zio.*
import com.alpha.domain.model.*
import java.util.UUID
import io.getquill.*

trait CategoryRepository:
  def findAll: Task[List[Category]]
  def findById(id: UUID): Task[Option[Category]]
  def findBySlug(slug: String): Task[Option[Category]]
  def findByParentId(parentId: UUID): Task[List[Category]]
  def findByIsActiveTrue: Task[List[Category]]
  def existsBySlug(slug: String): Task[Boolean]
  def findRootCategories: Task[List[Category]]
  def search(query: String): Task[List[Category]]
  def create(category: Category): Task[UUID]
  def update(category: Category): Task[Int]
  def delete(id: UUID): Task[Int]

object CategoryRepository:
  val layer: ZLayer[PostgresCtx, Nothing, CategoryRepository] =
    ZLayer.fromFunction(new CategoryRepositoryImpl(_))

class CategoryRepositoryImpl(ctx: PostgresCtx) extends CategoryRepository:

  import ctx.*

  override def findAll: Task[List[Category]] =
    ZIO.attempt:
      run(query[Category]).toList

  override def findById(id: UUID): Task[Option[Category]] =
    ZIO.attempt:
      run(query[Category].filter(_.id == lift(id))).headOption

  override def findBySlug(slug: String): Task[Option[Category]] =
    ZIO.attempt:
      run(query[Category].filter(_.slug == lift(slug))).headOption

  override def findByParentId(parentId: UUID): Task[List[Category]] =
    ZIO.attempt:
      run(query[Category].filter(_.parentId.exists(_ == lift(parentId)))).toList

  override def findByIsActiveTrue: Task[List[Category]] =
    ZIO.attempt:
      run(query[Category].filter(_.isActive == lift(true))).toList

  override def existsBySlug(slug: String): Task[Boolean] =
    ZIO.attempt:
      run(query[Category].filter(_.slug == lift(slug)).nonEmpty)

  override def findRootCategories: Task[List[Category]] =
    ZIO.attempt:
      run(query[Category].filter(_.parentId.isEmpty)).toList

  override def search(queryStr: String): Task[List[Category]] =
    ZIO.attempt:
      val pattern = "%" + queryStr + "%"
      run(query[Category]
        .filter(c => c.name.like(lift(pattern)) || c.description.exists(_.like(lift(pattern))))).toList

  override def create(category: Category): Task[UUID] =
    ZIO.attempt:
      run(query[Category].insertValue(lift(category)))
      category.id

  override def update(category: Category): Task[Int] =
    ZIO.attempt:
      run(query[Category]
        .filter(_.id == lift(category.id))
        .updateValue(lift(category)))
      1

  override def delete(id: UUID): Task[Int] =
    ZIO.attempt:
      run(query[Category].filter(_.id == lift(id)).delete)
      1
